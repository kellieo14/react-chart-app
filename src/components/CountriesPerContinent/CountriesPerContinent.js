import React, { useState, useEffect } from 'react';
import Doughnut from "../Graphs/Doughnut/Doughnut";

export const extractRegions = (regions, countries) => {

    const countriesPerRegion = (region) => {
        let area = region.toLowerCase();
        for (region in regions) {
            if (area === region) {
                regions[region]++
            }
        }
    }

    for (let country in countries) {
        let { region } = countries[country];
        switch (region) {
            case 'Europe':
                countriesPerRegion(region);
                break;
            case 'Americas':
                countriesPerRegion(region);
                break;
            case 'Oceania':
                countriesPerRegion(region);
                break;
            case 'Asia':
                countriesPerRegion(region);
                break;
            case 'Africa':
                countriesPerRegion(region);
                break;
            case 'Polar':
                countriesPerRegion(region);
                break;
            default:
                countriesPerRegion('other');
        }
    }
}

const CountriesPerContinent = (props) => {
    const countries = props.countries;
    const [data, setData] = useState({});

    useEffect(() => {
        const regions = ({
            europe: 0,
            americas: 0,
            asia: 0,
            oceania: 0,
            africa: 0,
            polar: 0,
            other: 0
        });

        extractRegions(regions, countries);
        setData({
            chartTitle: 'Countries per Continent',
            titles: ['Europe', 'Americas', 'Asia', 'Oceania', 'Africa', 'Polar', 'Other'],
            values: regions,
            colors: ['#0be2d7', '#a73e5c', '#0080d6', '#8dc63f', '#ffe119', '#ba59fa', '#fd7400'],
            position: 'left',
        })
    }, [countries]);

    return (
        <div className="">
            <div>
                <Doughnut data={data} />
            </div>
        </div>
    );
}

export default CountriesPerContinent;