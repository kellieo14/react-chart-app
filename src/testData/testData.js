export const countries = [
    {
        name: 'China', 
        region: 'Asia', 
        languages: [
            { iso639_1: "zh", iso639_2: "zho", name: "Chinese", nativeName: "中文 (Zhōngwén)"}
        ], 
        population: 1377422166,
        area: 9640011,
    }, 
    {
        name: 'India', 
        region: 'Asia',
        languages: [
            {iso639_1: "hi", iso639_2: "hin", name: "Hindi", nativeName: "हिन्दी"},
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 1295210000,
        area: 3287590,
    }, 
    {
        name: 'United States of America', 
        region: 'Americas', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 323947000,
        area: 9629091,
    },
    {
        name: 'Nigeria', 
        region: 'Africa', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 186988000,
        area: 923768,
    }, 
    {
        name: 'France', 
        region: 'Europe', 
        languages: [
            {iso639_1: "fr", iso639_2: "fra", name: "French", nativeName: "français"}
        ], 
        population: 66710000,
        area: 640679,
    }, 
    {
        name: 'Australia', 
        region: 'Oceania', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 24117360,
        area: 7692024,
    }, {
        name: 'Polarland', 
        region: 'Polar', 
        languages: [
            {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
        ], 
        population: 11239004,
        area: 1284000,
    }, {
        name: 'Otherland', 
        region: 'Other', 
        languages: [
            {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
        ], 
        population: 100,
        area: 619745,
    }
];