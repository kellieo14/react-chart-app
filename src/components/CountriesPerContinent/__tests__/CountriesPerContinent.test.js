import React from 'react';
import * as reactHooks from 'react';
import TestRenderer from 'react-test-renderer';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { countries } from '../../../testData/testData';
import CountriesPerContinent, { extractRegions } from '../CountriesPerContinent';
Enzyme.configure({adapter: new Adapter()});

describe('CountriesPerContinent', () => {
    let spyUseState;
    let spyUseEffect;
    let mockSetData;
    beforeEach(() => {
        spyUseState = jest.spyOn(reactHooks, 'useState')
            .mockImplementationOnce(() => [{}, mockSetData]);
        spyUseEffect = jest.spyOn(reactHooks, 'useEffect');
        mockSetData = jest.fn();
    });

    const regions = ({
        europe: 0,  
        americas: 0,
        asia: 0, 
        oceania: 0, 
        africa: 0, 
        polar: 0,
        other: 0
    });

    it('should match snapshot', () => {
        const testRenderer = TestRenderer.create(<CountriesPerContinent countries={countries}/>)
        expect(testRenderer.toJSON()).toMatchSnapshot();
    });

    it('should call useState and useEffect', () => {
        mount(<CountriesPerContinent countries={countries}/>)
        expect(spyUseState).toHaveBeenCalled();
        expect(spyUseEffect).toHaveBeenCalled();
        expect(mockSetData).toHaveBeenCalled();
    });

    it('should count countries per continent', () => {
        extractRegions(regions, countries);
        const expectedRegions = ({
            europe: 1,  
            americas: 1,
            asia: 2, 
            oceania: 1, 
            africa: 1, 
            polar: 1,
            other: 1
        });
        expect(regions).toEqual(expectedRegions);
    });
});
