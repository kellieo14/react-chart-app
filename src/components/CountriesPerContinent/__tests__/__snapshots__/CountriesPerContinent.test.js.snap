// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountriesPerContinent should match snapshot 1`] = `
<div
  className=""
>
  <div>
    <div
      className="doughnut"
    />
  </div>
</div>
`;
